File: 2337-move-pieces-to-obtain-a-string/2337-move-pieces-to-obtain-a-string.dart
class Solution {
  bool canChange(String start, String target) {
    int si = 0;
    for (int ti = 0; ti < start.length; ti++){
        // pass spaces from both start and target
        if (target[ti] == '_') continue;
        while (si < start.length && start[si] == '_') si++;
        // check valid
        if (si >= start.length) return false;
        if (target[ti] != start[si]) return false;
        if (target[ti] == 'L' && ti > si) return false;
        if (target[ti] == 'R' && ti < si) return false;
        si++;
    }
    for (si; si < start.length; si++){
        if (start[si] != '_') return false;
    }
    return true;
  }
}

File: 2337-move-pieces-to-obtain-a-string/2337-move-pieces-to-obtain-a-string.py
class Solution:
    def canChange(self, start: str, target: str) -> bool:
        l = len(start)
        si, ti = 0, 0
        while ti < l or si < l:
            while si < l and start[si] == '_':
                si += 1
            while ti < l and target[ti] == '_':
                ti += 1
            if si >= l and ti >= l:
                return True
            if si >= l or ti >= l:
                return False
            if target[ti] != start[si]:
                return False
            if start[si] == 'R' and ti < si:
                return False
            if start[si] == 'L' and ti > si:
                return False
            si += 1
            ti += 1
        return True
        